README.md
# Netatmo Indicator [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

Ubuntu debian package deployment of a Menubar indicator for the Netatmo Weather Station

## Installation
```
sudo add-apt-repository ppa:mihaigalos/ppa
sudo apt-get update
sudo apt install netatmo-indicator
netatmo-indicator &
```

## Screenshots
![alt text](screenshots/netatmo-indicator-screenshot.png)



## Settings
The indicator stores its settings in `~/.netatmo-indicator-preferences.yaml`, with an additional credentials file
for netatmo. The credentials' file is referenced with the <credentials_file> tag in ~/.netatmo-indicator-preferences.yaml

One may edit the credentials directly in the Menu.

If desired, one may edit the ~/.netatmo-indicator-preferences.yaml and add a dictionary of aliases to substitute the
names displayed. Make sure to close the indicator before you manually edit the file.

Example:
```
aliases:
  Living: In
  Outdoor: Out
  Bedroom: Bed
```

## A note on credentials
When the indicator first starts, the user is presented with the settings menu. One needs to fill in his netatmo username, password, cilentId and clientSecret. All this information can be gathered from the user's netatmo profile.

## Deployment [Self Note]
This is a self note. You do not need this for normal indicator usage.
##### Check if key already available
`gpg --list-secret-keys`

##### Generate key if not available
```
gpg --gen-key
gpg -a --output ~/.gnupg/mihaigalos.gpg --export 'Mihai Galos'
gpg --import ~/.gnupg/mihaigalos.gpg
```

##### Debian Package Generation
###### Generation for local usage / binary build
`dpkg-buildpackage -b -pgpg -kmihaigalos@gmail.com`

###### Generation for local deployment / source build
Launchpad (i.e.) only accepts source builds, and to avoid errors, we need an original version of the package see [askubuntu](https://askubuntu.com/questions/1087569/deploying-own-debian-package-to-launchpad) for further info.

The current folder contains i.e. `.git` which we do not want to deploy. We thus make a new folder `foo` and create the package there.

```
(in the git folder)
current_directory=$(basename "$PWD")

mkdir -p ../foo/${current_directory} && find . -mindepth 1 -maxdepth 1 -exec cp {*indicator,*.py,*.desktop,*.md} ../foo/${current_directory} \; && cp -r debian ../foo/${current_directory}
cd ../foo
tar -czvf ${current_directory}_0.1.orig.tar.gz ${current_directory}
cd ${current_directory}

dpkg-buildpackage -S -pgpg -kmihaigalos@gmail.com
```

##### Prepare for upload to Launchpad
###### Prerequisites:
* Ubuntu One account active
* gpg key used to sign the debian uploaded to Ubuntu Keyserver:
 - `gpg --list-keys`
 - `gpg  --keyserver hkp://keyserver.ubuntu.com:11371 --send-keys <KEY>`)
* OpenPGP keys displays the key

###### Upload source:
`dput ppa:mihaigalos/ppa *source.changes`


debian/changelog
netatmo-indicator (0.1-13) bionic; urgency=low

  * Initial release

 -- Mihai Galos <mihaigalos@gmail.com>  Tue, 23 Oct 2018 17:05:26 +0100


graph_generator.py
from collections import OrderedDict
import plotly, os
import plotly.plotly as py
import json
from plotly.graph_objs import *
import time

class MakePlot:
    def __init__(self):
        with open(os.path.join(os.path.expanduser('~'), 'netatmo_data.json')) as f:
            try:
                self.data = json.load(f)
                self.data = OrderedDict(sorted(self.data.items(), key=lambda t: t[0]))
            except Exception, err:
                print(err)
            f.close()

    def read_data(self):
        x= []
        y= [ [] for i in range(len(self.data.values()[1]))]

        captions = [key for (key, value) in self.data.values()[0].iteritems()]

        i = 0
        for k,v in self.data.iteritems():
            x.append(k.split(" ")[1])
            i=0
            for location, value in v.iteritems():
                y[i].append(value)
                i=i+1
        return (x, y, captions)

    def construct_trace(self, x, y, caption):
        return {
          "x": x,
          "y": y,
          "line": {"shape": "spline"},
          "mode": "lines+markers",
          "type": "scatter",
          "name": caption
        }

    def draw(self):

        x, y, captions = self.read_data()

        traces = []
        i = 0
        for y_value in y:
            trace = self.construct_trace(x, y_value, captions[i])
            traces.append(trace)
            i +=1

        data = Data(traces)

        layout = {
            "showlegend":False,
            "xaxis" : {"nticks":10}
        }
        fig = Figure(data=data, layout=layout)
        plot_url = plotly.offline.plot(fig)

MakePlot().draw()


netatmo-service-wrapper.py
#!/usr/bin/env python2

import requests
import yaml
import json


class Credentials:
    def read_credentials(self):

        try:
            with open("../netatmo-credentials.yaml", 'r') as stream:
                return yaml.load(stream)
        except yaml.YAMLError as exc:
            print(exc)


class Netatmo:

    def get_token(self):
        credentials = Credentials().read_credentials()

        response = requests.post('https://api.netatmo.net/oauth2/token',
                                 data={'grant_type' 	: 'password',
                                       'client_id'		: credentials["clientId"],
                                       'client_secret'	: credentials["clientSecret"],
                                       'username' 		: credentials["userId"],
                                       'password' 		: credentials["pass"],
                                       'scope'	: 'read_station'})

        access_token = eval(response.text)["access_token"]

        return access_token

    def get_modules(self, response):
        result = []
        modules = json.loads(response.text)["body"]["modules"]
        for module in modules:
            module_name = module["module_name"]
            try:
                temperature = module["dashboard_data"]["Temperature"]
                result.append([module_name, temperature])

            except:
                pass
        return {item[0]: item[1] for item in result}

    def get_devices(self, access_token):
        response = requests.post('https://api.netatmo.net//api/devicelist',
                                 data={'access_token' 	: access_token})
        modules = self.get_modules(response)
        print(modules)

        devices = {}
        return devices

    def get_data(self):
        access_token = self.get_token()
        devices = self.get_devices(access_token)


def main():
    netatmo = Netatmo()
    netatmo.get_data()


if __name__ == '__main__':
    main()


netatmo_service_wrapper.py
#!/usr/bin/env python2

import requests
import yaml
import json
from datetime import datetime

class Credentials:
    def make_default(self):
        result = {
            "clientId" : "",
            "clientSecret" : "",
            "email" : "",
            "password" : ""
        }
        return result

    def read_credentials(self, credentials_file):
        """ Read in yaml file containing the following format:
            clientId: <generated from netatmo. i.e. :abcde3f167e482a63d812345>
            clientSecret: <generated from netatmo. i.e. :abcdeDH06Fs7vL7Rd5j8R5O7zQW7TwvdOR3QQkV12345>
            userId: <yourmail@host.com>
            pass: <netatmo password>"""
        try:
            with open(credentials_file,'r') as stream:
                return yaml.load(stream)
        except yaml.YAMLError as exc:
            pass
            #print(exc)


class Netatmo:
    def __init__(self, credentials_file):
        self.credentials_file = credentials_file

    def get_token(self):
        credentials = Credentials().read_credentials(self.credentials_file)
        response = requests.post('https://api.netatmo.net/oauth2/token',
                                 data={'grant_type' 	: 'password',
                                       'client_id'		: credentials["clientId"],
                                       'client_secret'	: credentials["clientSecret"],
                                       'username' 		: credentials["email"],
                                       'password' 		: credentials["password"],
                                       'scope'	: 'read_station'})

        if response.ok != True:
            raise Exception(response.text)
        access_token = eval(response.text)["access_token"]
        return access_token

    def get_modules(self, response):
        result = []
        modules = json.loads(response.text)["body"]["modules"]

        for module in modules:
            module_name = module["module_name"]
            try:
                temperature = module["dashboard_data"]["Temperature"]
                result.append([module_name, temperature])

            except:
                pass
        return {item[0]: item[1] for item in result}

    def get_devices(self, response):
        result = {}
        devices = json.loads(response.text)["body"]["devices"]

        module_name = devices[0]["module_name"]
        temperature = devices[0]["dashboard_data"]["Temperature"]

        return {module_name: temperature}

    def get_timestamp(self, response):
        timestamp =""
        devices = json.loads(response.text)["body"]["devices"]
        raw_timestamp = int(str(devices[0]["dashboard_data"]["time_utc"]))
        timestamp = datetime.fromtimestamp(raw_timestamp).strftime('%Y-%m-%d %H:%M:%S')

        return timestamp

    def get_timestamped_modules_and_devices(self, access_token):
        response = requests.post('https://api.netatmo.net//api/devicelist',
                                 data={'access_token' 	: access_token})
        if response.ok != True:
            raise Exception(response.text)
        modules = self.get_modules(response)
        devices = self.get_devices(response)
        timestamp = self.get_timestamp(response)
        return (dict(modules, **devices), timestamp)

    def get_data(self):
        access_token = self.get_token()
        (devices, timestamp) = self.get_timestamped_modules_and_devices(access_token)
        return (devices, timestamp)


netatmo_indicator.py
#!/usr/bin/env python2

# -*- coding: utf-8 -*-

#
# Author: Mihai Galos
# Date: October 23 , 2018
# Purpose: indicator for the Netatmo Weather Station
# Tested on: Ubuntu 16.04 LTS
#
#
# Licensed under The MIT License (MIT).
# See included LICENSE file or the notice below.
#
# Copyright (c) 2018 Mihai Galos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
from netatmo_service_wrapper import Netatmo
import os
import json
import yaml

import gi
gi.require_version('Notify', '0.7')
gi.require_version('AppIndicator3', '0.1')
gi.require_version('Gtk', '3.0')

from gi.repository import Notify
from gi.repository import Gio
from gi.repository import Gtk as gtk
from gi.repository import AppIndicator3 as appindicator
from gi.repository import GLib as glib
from collections import namedtuple

from netatmo_service_wrapper import Credentials


class EntryWindow(gtk.Window):

    def __init__(self, credentials, prefs, callback_on_config_changed):
        gtk.Window.__init__(self, title="Settings")

        self.callback_on_config_changed = callback_on_config_changed

        self.metadata = {}
        self.metadata.update ( {"credentials" : credentials})
        self.metadata.update ( {"prefs" : prefs})
        self.metadata.update ({"ui": {}})

        self.set_border_width(6)
        self.set_default_size(300, 100)

        self.timeout_id = None

        self.vbox = gtk.Box(orientation=gtk.Orientation.VERTICAL, spacing=6)
        self.add(self.vbox)

        self.label = gtk.Label("Config file: ~/.netatmo-indicator-preferences.yaml")
        self.vbox.pack_start(self.label, True, True, 0)


        for k,v in credentials.iteritems():
            self.label_entry(k, v, self.vbox)

        for k,v in prefs.iteritems():
            if type(prefs[k]) is not dict:
                self.label_entry(k, v, self.vbox)

        hbox = gtk.Box(spacing=6)
        self.vbox.pack_start(hbox, True, True, 0)

        self.save_button = gtk.Button("Save")
        self.save_button.connect("pressed", self.on_save_button_pressed)
        hbox.pack_start(self.save_button, True, True, 0)

        self.cancel_button = gtk.Button("Cancel")
        self.cancel_button.connect("pressed", self.on_cancel_button_pressed)
        hbox.pack_start(self.cancel_button, True, True, 0)

        self.add(hbox)
        self.show_all()
        pass

    def label_entry(self, label_name, label_value, parent):
        hbox = gtk.Box(spacing=6)
        parent.add(hbox)

        entry = gtk.Entry()
        entry.set_width_chars(50)
        entry.set_text(label_value)
        hbox.pack_end(entry, False, False, 0)

        label = gtk.Label(label_name)
        hbox.pack_end(label, False, False, 0)

        self.metadata["ui"].update({label_name : entry})

    def on_save_button_pressed(self, button):

        data_to_write = {}

        credentials_file =self.metadata["ui"]["credentials_file"].get_text()
        data_to_write.update({"credentials_file": credentials_file})
        preferences_file = os.path.join(os.path.expanduser('~'), ".netatmo-indicator-preferences.yaml")

        credentials = {}

        for k,v in self.metadata["ui"].iteritems():
            if "credentials_file" != k:
                credentials.update({k : v.get_text()})


        with open(preferences_file, "w+") as outfile:
            yaml.dump(data_to_write, outfile, default_flow_style=False)

        with open(credentials_file, "w+") as outfile:
            yaml.dump(credentials, outfile, default_flow_style=False)

        self.hide()
        self.callback_on_config_changed()

    def on_cancel_button_pressed(self, button):
        self.hide()


class Menu:
    def __init__(self, app, netatmo, aliases, credentials, prefs, callback_on_config_changed):

        self.callback_on_config_changed = callback_on_config_changed

        if hasattr(self, 'app_menu'):
            for item in self.app_menu.get_children():
                self.app_menu.remove(item)
        self.app_menu = gtk.Menu()
        self.credentials = credentials
        self.prefs = prefs
        display_string = ""

        if netatmo.exception is None:
            degree_sign = u'\N{DEGREE SIGN}'
            for k, v in netatmo.data.iteritems():
                if k in aliases:
                    k = aliases[k]
                display_string = display_string+k+": "+str(v)+degree_sign+" "

            app.set_label(display_string, "")
        else:
            display_string = "Error communicating with Netatmo."
            app.set_label(display_string, "")
            contents = [self.app_menu, gtk.ImageMenuItem, 'help-hint',
                        str(netatmo.exception), self.quit, [None]]
            self.add_menu_item(*contents)


        contents = [self.app_menu, gtk.ImageMenuItem, 'gtk-info',
                    'Timestamp: '+netatmo.timestamp, None, [None]
                    ]
        self.add_menu_item(*contents)

        contents = [self.app_menu, gtk.ImageMenuItem, 'gtk-preferences',
                    'Settings', self.on_settings_clicked, [None]
                    ]
        self.add_menu_item(*contents)

        contents = [self.app_menu, gtk.ImageMenuItem, 'exit',
                    'Quit', self.quit, [None]
                    ]
        self.add_menu_item(*contents)

        self.app = app

    def on_settings_clicked(self, *args):
        entry_window = EntryWindow(self.credentials, self.prefs, self.on_config_changed)

    def add_menu_item(self, menu_obj, item_type, image, label, action, args):
        menu_item = self.create_menu_item(action, args, image, item_type, label)
        menu_obj.append(menu_item)
        menu_item.show()

    def create_menu_item(self, action, args, image, item_type, label):

        if item_type is gtk.ImageMenuItem and label:
            menu_item = self.create_menu_item_image_and_label(image, label)
        elif item_type is gtk.ImageMenuItem and not label:
            menu_item = self.create_menu_item_image(image)
        elif item_type is gtk.MenuItem:
            menu_item = gtk.MenuItem(label)
        elif item_type is gtk.SeparatorMenuItem:
            menu_item = gtk.SeparatorMenuItem()
        if action:
            menu_item.connect('activate', action, *args)
        return menu_item

    def create_menu_item_image(self, image):
        menu_item = gtk.ImageMenuItem()
        menu_item.set_always_show_image(True)
        if '/' in image:
            icon = gtk.Image.new_from_file(image)
        else:
            icon = gtk.Image.new_from_icon_name(image, 16)
        menu_item.set_image(icon)
        return menu_item

    def create_menu_item_image_and_label(self, image, label):
        menu_item = gtk.ImageMenuItem.new_with_label(label)
        menu_item.set_always_show_image(True)
        if '/' in image:
            icon = gtk.Image.new_from_file(image)
        else:
            icon = gtk.Image.new_from_icon_name(image, 48)
        menu_item.set_image(icon)
        return menu_item

    def get_menu(self):
        return self.app_menu

    def get_app(self):
        return self.app

    def quit(self, *args):
        gtk.main_quit()

    def on_config_changed(self):
        self.callback_on_config_changed()

class NetatmoIndicator(object):

    def __init__(self):
        current_folder =  os.getcwd()

        self.app = appindicator.Indicator.new(
            'netatmo-indicator', current_folder + "/FF4D00-0.0.png",
            appindicator.IndicatorCategory.SYSTEM_SERVICES
        )

        self.app.set_status(appindicator.IndicatorStatus.ACTIVE)

        self.user_home = os.path.expanduser('~')
        self.prefs_file = os.path.join(
            self.user_home, ".netatmo-indicator-preferences.yaml")

        self.note = Notify.Notification.new(__file__, None, None)
        Notify.init(__file__)

        self.cache_icon_theme = self.gsettings_get(
            'org.gnome.desktop.interface', None, 'icon-theme'
        )

        self.fetch_netatmo_data()

        self.make_menu()
        self.update()

    def fetch_netatmo_data(self):

        self.prefs = self.read_prefs_file()

        if "aliases" not in self.prefs:
            self.prefs.update({"aliases": {}})
        NetatmoContainer = namedtuple("NetatmoContainer", "data timestamp exception")
        try:
            self.credentials = Credentials().read_credentials(self.prefs["credentials_file"])
            netatmo = Netatmo(self.prefs["credentials_file"])
            (data, timestamp) = netatmo.get_data()
            exception = None
            self.netatmo = NetatmoContainer(data, timestamp, exception)

        except Exception, err:
            self.credentials={}
            data = {}
            timestamp="?"
            exception = err
            self.netatmo = NetatmoContainer(data, timestamp, exception)
            pass

    def run(self):
        try:
            gtk.main()
        except KeyboardInterrupt:
            pass

    def update(self):
        timeout = 5*60
        glib.timeout_add_seconds(timeout, self.callback)


    def callback(self):
        self.fetch_netatmo_data()

        self.make_menu()
        self.update()

    def make_menu(self):
        menu_object = Menu(self.app, self.netatmo, self.prefs["aliases"], self.credentials, self.prefs, self.callback)
        self.app = menu_object.get_app()
        self.app.set_menu(menu_object.get_menu())


    def gsettings_get(self, schema, path, key):
        """utility: get value of gsettings schema"""
        if path is None:
            gsettings = Gio.Settings.new(schema)
        else:
            gsettings = Gio.Settings.new_with_path(schema, path)
        return gsettings.get_value(key)

    def read_prefs_file(self, *args):

        if os.path.exists(self.prefs_file):
            try:
                with open(self.prefs_file) as f:
                    result = yaml.load(f)
                    f.close()
                    return result
            except yaml.YAMLError as exc:
                raise Exception("File does not exist: "+str(self.prefs_file))
        else:
            data_to_write = {"credentials_file" : os.path.join(self.user_home, ".netatmo-credentials.yaml"), "aliases" : {}}
            with open(self.prefs_file, "w") as outfile:
                yaml.dump(data_to_write, outfile, default_flow_style=False)

            with open(self.prefs_file) as infile:
                return yaml.load(infile)


    def write_prefs_file(self, *args):
        with open(self.prefs_file, 'w') as f:
            try:
                json.dump(self.prefs, f,indent=4, sort_keys=True )
            except Exception as e:
                self.send_notif(
                    self.note,
                    'Failed writing ' + self.prefs_file,
                    str(e)
                )

def main():
    indicator = NetatmoIndicator()
    indicator.run()


if __name__ == '__main__':
    main()
